### DISCUSSION TITLE
[Blazor] Dispose page before initializing the navigated page

### DISCUSSION THREAD

**Discussion by danielchalmers** [dotnet/aspnetcore#53535] [Q&A] [answered]
You have a page that implements IDisposable and you call NavigationManager.NavigateTo("/NewPage").

The OnInitialized() event on NewPage is called before Dispose() on the previous page.

Is it possible to reverse the order without forcing a reload?

**Comment by MariovanZeist** [comment]
Hi @danielchalmers 
You could add the ``NavigationLock`` component to your page, and handle the ``OnBeforeInternalNavigation`` event.
(Or call the ``NavigationManager.RegisterLocationChangingHandler(OnLocationChanging)`` function to register a function that gets called before the navigation takes place. (be sure to dispose it though). [See here] (https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-8.0#handleprevent-location-changes)

**Reply by danielchalmers** [reply]
Thank you, that works but it doesn't trigger on refresh which complicates things a little.

**Comment by javiercn** [comment, marked as answer]
It's not possible, we can only dispose the previous page/component when it is no longer part of the document.
